src/ripple/protocol/impl/STVector256.cpp

```diff
@@ -26,19 +26,19 @@ namespace ripple {
 
 STVector256::STVector256(SerialIter& sit, SField const& name) : STBase(name)
 {
-    Blob data = sit.getVL();
-    auto const count = data.size() / (256 / 8);
-    mValue.reserve(count);
-    Blob::iterator begin = data.begin();
-    unsigned int uStart = 0;
-    for (unsigned int i = 0; i != count; i++)
-    {
-        unsigned int uEnd = uStart + (256 / 8);
-        // This next line could be optimized to construct a default
-        // uint256 in the vector and then copy into it
-        mValue.push_back(uint256(Blob(begin + uStart, begin + uEnd)));
-        uStart = uEnd;
-    }
+    auto const slice = sit.getSlice(sit.getVLDataLength());
+
+    if (slice.size() % uint256::size() != 0)
+        Throw<std::runtime_error>(
+            "Bad serialization for STVector256: " +
+            std::to_string(slice.size()));
```

Comment:
The `NFTokenCancelOffer` uses a `STVector256` field. The old code may have accepted an incorrectly serialized `sfNFTTokenOffers` field where as the new code throws. I'd be surprised if an incorrectly serialized transaction could ever make into the ledger, but I'm not sure I can prove it (code like this would assert in debug mode: https://github.com/nbougalis/rippled/blob/c497a62fb95c98a1e29f0eb0557ca36adcb89c7d/src/ripple/app/tx/impl/Transactor.cpp#L785-L802)

I think it's fine as-is, but I did want to bring this up.

Reply:
Thanks for bringing this up. So two comments:

1. This cannot occur on the main network since the `featureNonFungibleTokensV1` amendment isn't active, and so the `NFTokenCancelOffer` transaction cannot be executed.
2. In addition to the code you showed, the code will call [`sterilize`](https://github.com/ripple/rippled/blob/25474343a98922ee5b44019f45657e381080954d/src/ripple/protocol/impl/STTx.cpp#L550-L557) on submitted transactions, which _should_ detect this in _some_ circumstances. 

Given than we're planning to introduce another fix amendment to correct a minor issue identified with the XLS-20 implementation, which (for obvious reasons!) should be activated _before_ XLS-20 implementation, I think my suggestion would be to include this code with this release. It doesn't need to be directly gated by the amendment, but if the fix amendment activates before `featureNonFungibleTokensV1` (and again, it should), then this issue will never be able to trigger.

Apropos of this, I wonder what the performance implications of _always_ executing the serialization/deserialization check that's currently only done if compiled with `-DDEBUG` would be. Also, I wonder whether this check more properly belongs in either `preflight` or `preclaim` (we ought to claim a fee if someone submits a broken transaction and entomb the transation in the ledger!).

There are good reasons to have it in `operator()` (i.e. during "application") too, but, as I said above, if this check fails, it would be better to `tec` out of there, instead of simply calling `assert`.